### DISCUSSION TITLE
Discriminated unions with wrap/before field validator that runs on all fields

### DISCUSSION THREAD

**Discussion by a3ng7n** [pydantic/pydantic#9906] [Question]
Hello - I'm trying run a `wrap` validator on every individual field, while also using one of those fields as a discriminator. I'm naturally getting the usage error that wrap/before can't be used on discriminator fields, which I understand the reasoning for - and I don't actually want this particular validator to run for discriminator fields (as I use them, which are typically just `Literal["banana"]`). Is there any way to achieve having a wrap validator run for every field *except* discriminator fields?

Approach A is what I'm trying to achieve, but notably throws `PydanticUserError: Cannot use a mode='wrap' validator in the discriminator field 'my_name' of Model 'ChildClassA'`:
```python filename="a.py"
from typing import Annotated, Any, Dict, Literal, Union

from pydantic import (
    BaseModel,
    Discriminator,
    Field,
    ValidationInfo,
    ValidatorFunctionWrapHandler,
    field_validator,
)

class ContextBase(BaseModel):
    @field_validator("*", mode="wrap")
    @classmethod
    def wrap_each_field(
        cls, v: Any, handler: ValidatorFunctionWrapHandler, info: ValidationInfo
    ) -> Any:
        # do something
        print(f"wrapping field: {info.field_name} of class: {cls.__qualname__}")
        result = handler(v)
        # do something else
        return result

class ChildClassA(ContextBase):
    my_name: Literal["ChildClassA"] = "ChildClassA"
    fruit: Annotated[str, Field(validate_default=True)] = "pear"

class ChildClassB(ContextBase):
    my_name: Literal["ChildClassB"] = "ChildClassB"
    amount: Annotated[int, Field(validate_default=True)] = 23

ChildrenClasses = Annotated[Union[ChildClassA, ChildClassB], Discriminator("my_name")]

class ContainerClass(ContextBase):
    children: Dict[str, ChildrenClasses]

# usage
data = {
    "children": {"abc": {"my_name": "ChildClassB"}, "123": {"my_name": "ChildClassA"}}
}

container = ContainerClass.model_validate(data)

assert type(container.children["abc"]) is ChildClassB
assert type(container.children["123"]) is ChildClassA

print(container)
```

Approach B hides the discrimination behind a `get_my_name` method, which seems to avoid throwing the usage error. I'm not a big fan since it requires declaring the name for `ChildClassX` in so many places. Also, I'm not positive but guessing that some type inferencing is degraded when looking at `ContainerClass.children`?:
```python filename="b.py"
from typing import Annotated, Any, Dict, Literal, Union

from pydantic import (
    BaseModel,
    Discriminator,
    Field,
    Tag,
    ValidationInfo,
    ValidatorFunctionWrapHandler,
    field_validator,
)

class ContextBase(BaseModel):
    @field_validator("*", mode="wrap")
    @classmethod
    def wrap_each_field(
        cls, v: Any, handler: ValidatorFunctionWrapHandler, info: ValidationInfo
    ) -> Any:
        # do something
        print(f"wrapping field: {info.field_name} of class: {cls.__qualname__}")
        result = handler(v)
        # do something else
        return result

class ChildClassA(ContextBase):
    my_name: Literal["ChildClassA"] = "ChildClassA"
    fruit: Annotated[str, Field(validate_default=True)] = "pear"

class ChildClassB(ContextBase):
    my_name: Literal["ChildClassB"] = "ChildClassB"
    amount: Annotated[int, Field(validate_default=True)] = 23

# function to "hide" accessing my_name as the discriminator
def get_my_name(v: Any) -> str:
    name = None
    if isinstance(v, dict):
        name = v.get("my_name")
    elif isinstance(v, BaseModel):
        name = v.model_dump().get("my_name")

    if name is None:
        raise ValueError("my_name not found!")

    return name

# need to use `Tag` annotations here to use function discriminator
ChildrenClasses = Annotated[
    Union[
        Annotated[ChildClassA, Tag("ChildClassA")],
        Annotated[ChildClassB, Tag("ChildClassB")],
    ],
    Discriminator(get_my_name),
]

class ContainerClass(ContextBase):
    children: Dict[str, ChildrenClasses]
```

Approach C does some hacky stuff with parts of pydantic I don't fully understand, but seems to work:
```python filename="c.py"
from typing import Annotated, Any, Dict, Literal, Union, get_origin

from pydantic import (
    BaseModel,
    Discriminator,
    Field,
    ValidationInfo,
    ValidatorFunctionWrapHandler,
    field_validator,
)

from pydantic._internal._decorators import FieldValidatorDecoratorInfo, Decorator

class ContextBase(BaseModel):
    # put this here to give field_validator an existing field to match against
    placeholder_field: Any = None

    # manually modify which fields the `wrap_each_field` runs for
    @classmethod
    def __pydantic_init_subclass__(cls, **_):
        method_name = "wrap_each_field"
        fields_to_validate = []

        for field_name, field in cls.model_fields.items():
            # Just ignore validation of *any* Literal
            if get_origin(field.annotation) == Literal:
                continue

            # Ignore validation of the placeholder field
            if field_name == "placeholder_field":
                continue

            fields_to_validate += [field_name]

        fields_to_validate = tuple(fields_to_validate)

        info = FieldValidatorDecoratorInfo(
            fields=fields_to_validate, mode="wrap", check_fields=None
        )
        deco = Decorator.build(cls, cls_var_name=method_name, info=info, shim=None)
        cls.__pydantic_decorators__.field_validators[method_name] = deco

        cls.model_rebuild(force=True)
        return super().__pydantic_init_subclass__()

    # "placeholder_field" here gets replaced in __pydantic_init_subclass__
    @field_validator("placeholder_field", mode="wrap")
    @classmethod
    def wrap_each_field(
        cls, v: Any, handler: ValidatorFunctionWrapHandler, info: ValidationInfo
    ) -> Any:
        # do something
        print(f"wrapping field: {info.field_name} of class: {cls.__qualname__}")
        result = handler(v)
        # do something else
        return result

class ChildClassA(ContextBase):
    my_name: Literal["ChildClassA"] = "ChildClassA"
    fruit: Annotated[str, Field(validate_default=True)] = "pear"

class ChildClassB(ContextBase):
    my_name: Literal["ChildClassB"] = "ChildClassB"
    amount: Annotated[int, Field(validate_default=True)] = 23

ChildrenClasses = Annotated[Union[ChildClassA, ChildClassB], Discriminator("my_name")]

class ContainerClass(ContextBase):
    children: Dict[str, ChildrenClasses]
```

tl;dr: the third approach, approach C, is the one way I've come up with to get what I want, but would like to know:

1. is it even correct usage of `__pydantic_decorators__`, or is there some issue or future change that'll make this not work? and/or..
2. is there a way to do this that's less hacky/more in-line with the public interfaces of pydantic?
